File: datatype/tap_port.go
package datatype

import (
	"fmt"

	"gitlab.yunshan.net/yunshan/droplet-libs/utils"
)

const (
	TAPPORT_FROM_MAC = iota
	TAPPORT_FROM_IPV4
	TAPPORT_FROM_IPV6
	TAPPORT_FROM_ID // 专属类型采集器时使用dispatcher id
	TAPPORT_FROM_NETFLOW
	TAPPORT_FROM_SFLOW
)

const (
	_FROM_OFFSET = 60
)

// 64     60                                    0
// +------+-------------------------------------+
// | from |              ip/id/mac              |
// +------+-------------------------------------+
type TapPort uint64

func FromMAC(mac uint32) TapPort {
	return TapPort(mac) | TAPPORT_FROM_MAC<<_FROM_OFFSET
}

func FromNetFlow(mac uint32) TapPort {
	return TapPort(mac) | TAPPORT_FROM_NETFLOW<<_FROM_OFFSET
}

func FromSFlow(mac uint32) TapPort {
	return TapPort(mac) | TAPPORT_FROM_SFLOW<<_FROM_OFFSET
}

func FromIP(ip uint32, isIPv6 bool) TapPort {
	tapPort := TapPort(ip)
	if !isIPv6 {
		tapPort |= TAPPORT_FROM_IPV4 << _FROM_OFFSET
	} else {
		tapPort |= TAPPORT_FROM_IPV6 << _FROM_OFFSET
	}
	return tapPort
}

func FromID(id int) TapPort {
	return TapPort(id) | TAPPORT_FROM_ID<<_FROM_OFFSET
}

func (p TapPort) SplitToPortAndType() (uint32, uint8) {
	return uint32(p & 0xffffffff), uint8(p >> _FROM_OFFSET)
}

func (p TapPort) String() string {
	tapPort, tapPortType := p.SplitToPortAndType()
	switch tapPortType {
	case TAPPORT_FROM_MAC:
		return fmt.Sprintf("MAC@%02x:%02x:%02x:%02x",
			uint8(tapPort>>24), uint8(tapPort>>16), uint8(tapPort>>8), uint8(tapPort))
	case TAPPORT_FROM_IPV4:
		return fmt.Sprintf("IPv4@%s", utils.IpFromUint32(tapPort))
	case TAPPORT_FROM_IPV6:
		return fmt.Sprintf("IPv6@0x%08x", tapPort)
	case TAPPORT_FROM_ID:
		return fmt.Sprintf("ID@%d", tapPort)
	case TAPPORT_FROM_NETFLOW:
		return fmt.Sprintf("NetFlow@%d", tapPort)
	case TAPPORT_FROM_SFLOW:
		return fmt.Sprintf("SFlow@%d", tapPort)
	default:
		panic(fmt.Sprintf("Invalid TapPort type is %d.", tapPortType))
	}
}

File: src/common/tap_port.rs
use std::fmt;
use std::net::Ipv4Addr;

// 64     60                                    0
// +------+-------------------------------------+
// | from |              ip/id/mac              |
// +------+-------------------------------------+
#[derive(Default)]
pub struct TapPort(u64);

impl TapPort {
    pub const FROM_LOCAL_MAC: u8 = 0;
    pub const FROM_GATEWAY_MAC: u8 = 1;
    pub const FROM_TUNNEL_IPV4: u8 = 2;
    pub const FROM_TUNNEL_IPV6: u8 = 3;
    pub const FROM_ID: u8 = 4;
    pub const FROM_NETFLOW: u8 = 5;
    pub const FROM_SFLOW: u8 = 6;
    const FROM_OFFSET: u64 = 60;

    pub fn from_local_mac(mac: u32) -> TapPort {
        TapPort(mac as u64 | ((TapPort::FROM_LOCAL_MAC as u64) << TapPort::FROM_OFFSET))
    }

    pub fn from_netflow(mac: u32) -> TapPort {
        TapPort(mac as u64 | ((TapPort::FROM_NETFLOW as u64) << TapPort::FROM_OFFSET))
    }

    pub fn from_sflow(mac: u32) -> TapPort {
        TapPort(mac as u64 | ((TapPort::FROM_SFLOW as u64) << TapPort::FROM_OFFSET))
    }

    pub fn from_gateway_mac(mac: u32) -> TapPort {
        TapPort(mac as u64 | ((TapPort::FROM_GATEWAY_MAC as u64) << TapPort::FROM_OFFSET))
    }

    pub fn from_tunnel_ip(ip: u32, is_ip_v6: bool) -> TapPort {
        TapPort(
            ip as u64
                | ((if is_ip_v6 {
                    TapPort::FROM_TUNNEL_IPV6
                } else {
                    TapPort::FROM_TUNNEL_IPV4
                } as u64)
                    << TapPort::FROM_OFFSET),
        )
    }

    pub fn from_id(id: u32) -> TapPort {
        TapPort(id as u64 | ((TapPort::FROM_ID as u64) << TapPort::FROM_OFFSET))
    }

    pub fn split_to_port_and_type(&self) -> (u32, u8) {
        (self.0 as u32, (self.0 >> TapPort::FROM_OFFSET) as u8)
    }
}

impl fmt::Display for TapPort {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let (p, t) = self.split_to_port_and_type();
        match t {
            TapPort::FROM_LOCAL_MAC => {
                let bs = p.to_be_bytes();
                write!(
                    f,
                    "LMAC@{:02x}:{:02x}:{:02x}:{:02x}",
                    bs[0], bs[1], bs[2], bs[3]
                )
            }
            TapPort::FROM_GATEWAY_MAC => {
                let bs = p.to_be_bytes();
                write!(
                    f,
                    "GMAC@{:02x}:{:02x}:{:02x}:{:02x}",
                    bs[0], bs[1], bs[2], bs[3]
                )
            }
            TapPort::FROM_TUNNEL_IPV4 => {
                write!(f, "IPv4@{}", Ipv4Addr::from(p))
            }
            TapPort::FROM_TUNNEL_IPV6 => {
                write!(f, "IPv6@{:#10x}", p)
            }
            TapPort::FROM_ID => {
                write!(f, "ID@{}", p)
            }
            TapPort::FROM_NETFLOW => {
                write!(f, "NetFlow@{}", p)
            }
            TapPort::FROM_SFLOW => {
                write!(f, "sFlow@{}", p)
            }
            _ => panic!("Invalid tap_port type {}.", t),
        }
    }
}

impl fmt::Debug for TapPort {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self)
    }
}
